Added additional File Header fields for version, copyright, and custo…

deanranderson opened the pull request:
…m lines

# Description
I've been enjoying using doxdocgen in VS Code, but I wanted a bit more flexibility to define my source file's Doxygen blocks.  So I added a couple new tags/templates.  It might be a bit rough in its current form!

## Feature

- [ ] Added additional options for File Doxygen block
  -  @version tag
  - @copyright tag, as a string array to allow multiple lines (place @n on each line to create line breaks)
  - copyrightTag will generate {year} templates
  - "customTag", array of strings for appending additional, custom lines
  - customTag will generate {date} templates too

A quick example header block with the additional fields:
![image](https://user-images.githubusercontent.com/41972662/46106430-a31b3000-c1a6-11e8-9416-55d428b22559.png)

christophschlosser commented on src/Lang/Cpp/CppDocGen.ts:
```diff
@@ -228,6 +228,34 @@ export class CppDocGen implements IDocGen {
+        this.cfg.File.customTag.forEach((element) => {
+            this.generateFromTemplate(
+                lines,
+                this.cfg.dateTemplateReplace,
```
Please check if this isn't set to an empty string.

deanranderson replied:
I want to allow empty lines in the customTag.  If the array is completely empty, the for loop shouldn't have any iterations.  For instance, the following should be a valid block to add:
```json
[ "@par First Paragraph",
"First Paragraph Text",
""
"@par Second Paragraph",
"Second Paragraph Text"
]

christophschlosser replied:
Sorry I wasn't clear on that.
I wanted you to check if `this.cfg.dateTemplateReplace` was empty as it can be set by the user and then the replacement would fuck up. Take a look at line #260
Your solution otherwise is fine.

deanranderson replied:
Ah, of course.  That makes sense.

christophschlosser commented on src/Lang/Cpp/CppDocGen.ts:
```diff
@@ -228,6 +228,34 @@ export class CppDocGen implements IDocGen {
+                lines,
+                this.cfg.dateTemplateReplace,
+                element,
+                [moment().format(this.cfg.Generic.dateFormat)],
```
What about other tags like `@author` or `@file`?
Maybe this needs another switch case where different keywords can be handled.

deanranderson replied:
That's a great feature, I'll see if I can figure out how to  substitute multiple sub-strings within a template.  It's pretty common for us to use a "History" section in these headers, which would have lines like: ```* {date} {author} created file.```  (Have not used typescript before yesterday)

I'd love to figure out how to query the author name and e-mail from Git.

christophschlosser replied:
You're doing an awesome job with typescript here then!
If you want to interact with Git maybe take a look at GitLens implementation:
https://github.com/eamodio/vscode-gitlens/blob/14e2aa67bfcd432a6b5af297cd5c20aafc182600/src/gitService.ts#L934
They have something to query the git user details for the current repo.
If you need some help you can feel free to ask!

christophschlosser commented:
Thank you for your work! :)
This is a great addition and I would love to have it in the addon.

deanranderson commented:
Added some tests, and fixed my linter issues.  Instead of bailing on empty date format, I default to use ISO in the custom tag.  I suppose there's the possibility the format is malformed too, but that isn't handled by the date tag currently, either.
Let me know if you'd like further refinements.  Cheers!

christophschlosser commented on package.json:
```diff
@@ -126,10 +141,20 @@
           "description": "Set the style of the author tag and your name.",
           "type": "string",
-          "default": "@author your name"
+          "default": "@author {author} ({email})"
```
Please use a new tag for this. I don't want it to break existing configs.

deanranderson replied:
Does it break existing configs?  Users can always hardcode their name in the authorTag, rather than using the {author}/{email} templates.

christophschlosser replied:
Yeah you're right. Just add it to the description which tags are supported.

christophschlosser commented on src/Lang/Cpp/CppDocGen.ts:
```diff
@@ -228,6 +247,44 @@ export class CppDocGen implements IDocGen {
+        }
+        // For each line of the customTag
+        this.cfg.File.customTag.forEach((element) => {
+            // Allow any of date, year, author, email to be replaced
```
This should also be in the description of the custom tag for the settings file.
Maybe some sort of list of which replacements are supported for the custom block.

christophschlosser commented on src/Lang/Cpp/CppDocGen.ts:
```diff
@@ -105,6 +105,18 @@ export class CppDocGen implements IDocGen {
+            if (i < param.length) {
+              template = template.replace(replace[i], param[i]);
+            }
+            // TODO: warn if mismatch string lengths?  Probably should use tuple of tuples
```
What do you mean by that?

deanranderson replied:
I'm just a bit wary of passing two different sized arrays to a function as different arguments.  A single argument with an array of [string,string] tuples might be safer (or at least caught at compile time?)

christophschlosser replied:
Yep. Good thinking!
